frontend/catalyst/compilation_pipelines.py

```diff
@@ -688,13 +768,30 @@ def get_cmain(self, *args):
         return function.get_cmain(*args)
 
     def __call__(self, *args, **kwargs):
+        static_argnums = self.compile_options.static_argnums
+
         if EvaluationContext.is_tracing():
             return self.user_function(*args, **kwargs)
 
         function, args = self._ensure_real_arguments_and_formal_parameters_are_compatible(
             self.compiled_function, *args
         )
-        recompilation_needed = function != self.compiled_function
+
+        recompilation_needed = (
+            function != self.compiled_function
+        ) and function not in self.stored_compiled_functions.values()
+
+        # Check if a function is created and add newly created ones into the hash table.
+        if static_argnums and recompilation_needed:
+            static_args_hash = self.get_static_args_hash(*args)
+            workspace = self.workspace
+            self.stored_compiled_functions[static_args_hash] = (function, workspace)
+            # Create new space for the next function to avoid using the spaces from
+            # the previously compiled ones.
+            self.workspace = WorkspaceManager.get_or_create_workspace(
+                self.function_name, self.preferred_workspace_dir
+            )
```

Comment:
This part is a bit strange. I know where you're coming from since this sets up the workspace for a future recompilation, but it also means the workspace property of the `QJIT` object just points to an empty directory after the function was compiled, which is not what someone using this code (and wanting to inspect the temporary files say) might expect. 

I think this property needs to be reset only when needed, i.e. before recompilation is about to happen.

Reply:
I see. I have moved this part into ``_ensure_real_arguments_and_formal_parameters_are_compatible`` so that the workspace creation happens before recompilation.